### ISSUE TITLE
privacy.resistFingerprinting.letterboxing.dimensions not working right on max vertical resolution

### ISSUE THREAD

**Issue by quark2323** [arkenfox/user.js#1214]
user_pref("privacy.resistFingerprinting", true);
With this option turned on, the reported resolution is not that of the display itself, rather of the rendered website resolution. Taking my view options into account (bookmarks bar, tab bar etc) the vertical viewable distance of the website is 1239 pixels instead of 1440.
I tried manually adding:
user_pref("privacy.resistFingerprinting.letterboxing", true);
user_pref("privacy.resistFingerprinting.letterboxing.dimensions", "2560x1440"); but doesn't seem to work, reported resolution is still 2560x1239. I think I tried to force a lower vertical resolution and it worked, so the option doesn't work for sizes higher than the rendered website.
If I turn privacy.resistFingerprinting off then I get 2560x1440, which is more common than 2560x1239.

**Comment by Thorin-Oakenpants** [comment]
NOTE: the letterboxing dimensions pref is unsupported

I'm having a hard time following what you are doing, and some of it doesn't make sense. Use max if you mean maximized, full-screen if you mean full-screen.

It has nothing to do with any website - it's all about your own device: devicePixelRatio, screen resolution, `layout.css.devPixelsPerPx`, system scaling setting, and even maybe some bugs on initial new window sizes, such as tiling managers, what extra chrome elements are showing e.g. toolbar/ menubar / sidebar, can be affected by themes and densities .. etc. And the values are affected by zoom.

So...
- What is your real screen size?
- With RFP, firefox cannot open maximized or full screen, and tries to open so the inner window is 1000x1000 at most, and fits your screen (i.e not bigger than visually allowable), in stepped ranges - these sizes match letterboxing
- If you want to start Firefox in a larger size, you can do that with the two prefs in `4502` but the values are still stepped regardless of what you put in
- if you enable letterboxing, then that too is in stepped ranges

So, you are trying to control the reported screen size by trying to get your inner window to always be 2560x1440 because somehow you think this protects you. In order to do that, how do you start Firefox, at what size - given that all starting sizes are rounded - do you maximize or something? Don't get me wrong, I'm trying to understand what you are doing here, and you can use an extension to resize the browser larger than your actual screen and position it anywhere - e.g. if you dragged a Firefox window from a higher res screen to a lower res secondary screen.

What is your actual screen res? It won't work if you can't physically fit 2560x1440 in a maximized or full screen (states which do not go off screen)?

As for the LB dimensions pref, that isn't controlled by any hardcoded stepped ranges, and I have tested it before, but not for a long time.

FWIW I have a 2560x1440 monitor, I start Firefox at 1600x900 and I do nothing special with letterboxing (which is on), so if I resize, or maximize or go full screen, it goes to work and the inner window is rounded

If the allowed inner windows letterbox sizes were say `1500x800, 1500x900, 1500x1000, 1600x800, 1600x900, 1600x1000, 1700x800, 1700x900, 1700x1000, 2560x1440` then my start up protection would work: if I accidentally open a little to big or small or accidentally resize .. then the sizes around it would cover me, and if I go full screen, then it should also work - i.e zero chrome issues, my actual screen res is 2560x1440 and thus the letterbox dimensions will fit. I might test it later if I get time, but I'm in no hurry if I can't understand STR

Maximizing != full screen. Can you be clearer in what you are doing, thanks

**Comment by quark2323** [comment]
My full resolution is 2560x1440. I'd like the browser to always report it as my resolution. Ideally it would not report any resizing.
I'm not talking about browser fullscreen mode, just maximizing the window.
About the fact that I think it protects me or not, it's just less info that my browser is giving out. Having it stick to reporting a certain resolution is overall less info. Right now, running the browser window maximized makes for extra bits of info the browser gives away.
Any way of making it always report a certain resolution?

**Comment by quark2323** [comment]
Here's my overrides:
`user_pref("browser.startup.page", 3); // 0102
user_pref("browser.search.suggest.enabled", true);
user_pref("browser.urlbar.suggest.searches", true);
user_pref("permissions.default.camera", 2);
user_pref("permissions.default.microphone", 2);
user_pref("dom.event.clipboardevents.enabled", false);
user_pref("browser.download.folderList", 2);
user_pref("browser.download.forbid_open_with", true);
user_pref("privacy.clearOnShutdown.history", false); // 2803
user_pref("privacy.cpd.history", false); // 2804 to match when you use Ctrl-Shift-Del
user_pref("accessibility.typeaheadfind", true); // enable "Find As You Type"
user_pref("browser.backspace_action", 0); // 0=previous page, 1=scroll up, 2=do nothing
user_pref("browser.quitShortcut.disabled", true); // disable Ctrl-Q quit shortcut [LINUX] [MAC]
user_pref("browser.tabs.loadBookmarksInTabs", true); // open bookmarks in a new tab [FF57+]
user_pref("browser.urlbar.decodeURLsOnCopy", true); // see bugzilla 1320061 [FF53+]
user_pref("extensions.pocket.enabled", false); // Pocket Account [FF46+]
user_pref("identity.fxaccounts.enabled", false); // Firefox Accounts & Sync [FF60+] [RESTART]
user_pref("reader.parse-on-load.enabled", false); // Reader View
user_pref("network.manage-offline-status", false); // see bugzilla 620472

user_pref("privacy.resistFingerprinting", false);`

Run the EFF test: https://coveryourtracks.eff.org/
Search for "Screen Size and Color Depth" in the results. Now switch privacy.resistFingerprinting to true and run the test again, and check on the detected resolution. Have a random size window for this test.

**Comment by Thorin-Oakenpants** [comment]
> My full resolution is 2560x1440. I'd like the browser to always report it as my resolution. Ideally it would not report any resizing.

If you use RFP then the screen is reported as the inner window. This is not randomized but rather tied to your inner window. YOu cannot make it NOT report the resizing using prefs. RFP means screen is tied to inner window, period, end of story.

> I'm not talking about browser fullscreen mode, just maximizing the window.

Maximizing cannot use your full screen, chrome elements prevent it. That's what full-screen is for.

> it's just less info that my browser is giving out

It's not less info. You're still reporting a value. If you mean it's a different value that is more common, that might be the case individually for that one fingerprinting metric, but when you combine it with others you will become unique

all RFP users, the screen = 100's x 200's or whatever unless they resize to weird sizes
all RFP users + letterboxing, the screen = 100's x 200's regardless
 - e.g. 2560x1440 users at FS = 2400 x 1400 = one of those 32 results below
 - e.g. 2560x1440 maximized = 2400 x 1200 (for me, I have toolbar enabled, but also a double height task bar) = one of those 32 results below

basically, the number of possible sizes reported with letterboxing drops.
- after 1000x1000, the sizes possible in a 2560x1440 are width 1560 x height = 440 = 686+ thousand results
- with letterboxing that drops to widths in 200s and heights in 100s = 8 x 4 = 32 results

Anything outside expected RFP new window sizes and letterboxed values = most likely unique, when paired with the fact you are using RFP, which can be 100% reliably detected in less than 1ms

> Run the EFF test: 

Read the [wiki](https://github.com/arkenfox/user.js/wiki/4.1-Extensions#small_orange_diamond-%EF%B8%8F-anti-fingerprinting-extensions-fk-no) .. scroll down to the bit that says stop using entropy figures from these sites, they're all BS because that's NOT how FPing works, i.e by individual metrics.


tl;dr: don't fiddle with RFP protected metrics

**Comment by quark2323** [comment]
Yes but RFP leaks extra info tho, on the screen resolution part. Without it, there's only 2560x1440 reported, irrespective of window sizes.
Using RFP offers more info into what I'm doing, vs without it, from the resolution point of view. The title of this issue is wrong as I didn't correctly understand what letterboxing does. 
So, using RFP there's no way I can stop the extra info leak regarding resolution?

**Comment by quark2323** [comment]
regarding the wiki and eff I don't see how it is inferior to the subject at hand. I used it, and gave it as an example, so you can see the issue with the resolution. like it or not, it's a tool that offers info on various bits, in this case the resolution info that my browser makes available to websites.

**Comment by Thorin-Oakenpants** [comment]
RFP doesn't "leak". 

Look, for better or worse, RFP, designed for Tor browser, decided, many years ago, to tie `inner` window measurements to `outer`, `available screen` and `screen`, and to start everyone off as `1000x1000` or smaller stepped sizes if it couldn't fit. So most users would be a square (partly to help hide orientation, and partly because of 4/3 screen aspect ratios)

Personally, I don't think screen + available screen should be tied to inner.
- that doesn't mean I don;t think screen shouldn't be protected
- first of all, scripts don't bother using inner measurements for FPing because they're not stable : users can resize, zoom, maximize, go full-screen etc. Maximized and full screen would be semi-stable, but it's just not really a stable metric to collect to link traffic. Note that zoom affects screen measurements anyway, as well as other factors like system scaling.
- secondly, as you say, tying screen to inner, means now the user has to control the inner window in order to not differ from the others - so all TB users are told (in comments, I have yet to see anything official in writing) .. don't resize, don't maximize, don't go full-screen, don't toggle the sidebar on, don't toggle the appmenu, don't toggle the toolbar .. blah blah blah - chrome elements fuck things up, because the sizing is only run when creating a NEW window
- so all up, the solution, from many many years ago, actually created some "problems" that didn't exist before: namely anything that changed the inner window, also changed the screen. Don't get me wrong, it also solved a lot of problems

So letterboxing was created, about two years ago, to at least severely limit the combos that inner returns - this is good. And can still be used to limit inner returns, regardless of screen
- But people still don't or shouldn't resize because this has been ingrained into them for ages (e.g. in TB), and the thinking is that most people don't resize, so you shouldn't either to hide in the masses. But if you do resize, at least the combos are much less, especially if you only resize now and then for some sites - e.g. go FS to watch a video, maximize to read an article properly, whatever
- But this is a usability issue - being stuck as 1000 max width.
   - New window sizes and letterboxing are desktop only (i.e not android). Hiding orientation is a waste of time and not even fully covered, and is a probably a waste of time. Tablets present an interesting cross, but normally the orientation is static (desktop), and on-demand (mobiles), who cares - its not stable and its simple to detect desktop vs mobile, etc. A res is a res regardless of `widthxheight` vs `heightxwidth`. This whole square thing is a bit moot. Desktops for a long time now, are nothing less than 16/9 or thereabouts,
   - At a max 1000 wide, even internal Firefox/Tor Browser pages end up creating horizontal scroll bars. Lots of websites seem to use 1024px as a point to determine whether to serve mobile web pages. All that width just going to waste to possibly help hide a few old screens at 1024x768 .. and wait .. they are already different because of the height limitation, i,.e they end up at 1000x700 or a few variants - they already stick out.

Allowing a better default max width, i.e landscaped, then usability is improved. Then by uncoupling the screen from inner (but still protecting the sceen, more on that in a sec), the inner window size does not have to be protected tooth and nail. Users can still stay at the default NEW window size, (e.g. 1000x1400) and letterboxing can still protect combos in resizing,
- The screen can simply be related to the inner, but stepped in some common resolutions - see CB info below, which came from a suggestion by me

This would improve usability / real-estate and actually reduce screen entropy (e.g. four or five results). The screen entropy would be mutable helping to thwart linkability. And it would not affect inner entropy, at all - you still have new window sizes and letterboxing. But you have to let RFP do it. In the meantime, CB added this on my recommendation (I wanted something to illustrate the point to TB)
- e.g as you inner window changes, the screen res reported goes up in steps

I have been planning for a year or more to push TB on this with some data and whatif scenarios, but the simple fact is that the aspect ratio cannot change, the only variables are scaling. So a wider default width is simple. And decoupling screen from inner is already mathematically proven = e.g. 32 buckets above 1000x1000 alone on a 2560x1440, not to mention those sizes under 1000x1000 < 5 screen buckets. Sheesh, even if all inners under 1000x1000 just reported 1600x900, that alone reduces the screen entropy for all the variations in new window sizes.  I'm not sure what I need to do to actually push this change, except get off my ass and make the case

Anyway,.. if you really want to do what you want to, just use CB's screen protection - thanks to ME and kkapsner. Be aware though, that your real screen value is returned in pure css (due to extension limitations) - and that can be read by JS, e.g. pseudo attributes

> like it or not, it's a tool that offers info on various bits

The wiki says nothing about sites like EFF except that the ENTROPY figures are not real world or reliable. Quit the strawman argument that the values it returns are not informative, **_which no one said, ever_**.

**Comment by quark2323** [comment]
I don't understand why we discuss about it (eff) since it has no value for the subject of discussion.
Thanks for explaining what's what.
